sematic/abstract_future.py

```diff
@@ -257,3 +310,36 @@ def is_root_future(self):
 
 def make_future_id() -> str:
     return uuid.uuid4().hex
+
+
+def get_minimum_call_chain_timeout_seconds(
+    future: AbstractFuture,
+) -> Tuple[Optional[float], Optional[AbstractFuture]]:
+    """Given a future, determine the most restrictive timeout up its call chain.
+
+    Parameters
+    ----------
+    future:
+        The future to start traversal at.
+
+    Returns
+    -------
+    If no futures up the call chain for this future have active timeouts,
+    returns (None, None). Otherwise returns a tuple where the first element
+    is the time until the most restrictive timeout expires and
+    the second element is the future that had specified that timeout.
+    """
+    call_chain = [future]
+    while call_chain[-1].parent_future is not None:
+        call_chain.append(call_chain[-1].parent_future)
+
+    remaining_time_future_pairs = [
+        (future.props.remaining_timeout_seconds, future)
+        for future in call_chain
+        if future.props.remaining_timeout_seconds is not None
+    ]
```

Comment:
Can these two loops be combined into one?

Reply:
Yes, but I don't think it would necessarily be easier to read. I think it's a matter of taste 🤷